base/lthooks.dtx

```diff
@@ -1970,54 +1995,197 @@
+%       following results:
+%     \begin{itemize}
+%       \item [@@_if_structure_exist:nTF]      \emph{unchanged}.
+%       \item [@@_if_usable:nTF]   returns |false|.
+%       \item [@@_if_declared:nTF]  returns |true|.
+%       \item [@@_if_disabled:nTF]  returns |true|.
+%     \end{itemize}
+%       The  structure test is unchanged (if the hook was unknown before it is
+%       false, otherwise true). The usable test returns false so that
+%       any \cs{UseHook} will bypass the hook from now on. The
+%       declared test returns true so that any further \cs{NewHook}
+%       generates an error and the disabled test returns true so that
+%       \cs{AddToHook} can return an error.
+% \fmiinline{maybe it should do this only after begin document?}
 %
 %    \end{description}
-%  \end{macro}
 %
 %
 %
 %
-%  \begin{macro}{\hook_new:n}
+% \subsubsection{Setting hooks up}
+%
+%
+%  \begin{macro}{\hook_new:n,\@@_new:n}
 %    The \cs{hook_new:n} declaration declare a new hook and expects
 %    the hook \meta{name} as its argument, e.g.,
 %    \hook{begindocument}.
 %    \begin{macrocode}
 \cs_new_protected:Npn \hook_new:n #1
   { \@@_normalize_hook_args:Nn \@@_new:n {#1} }
+%    \end{macrocode}
+%
+%    \begin{macrocode}
 \cs_new_protected:Npn \@@_new:n #1
   {
 %    \end{macrocode}
-%    We check for one of the internal data structures and if it
-%    already exists we complain.
+%   We check if the hook was already \emph{explicitly} declared with
+%   \cs{hook_new:n}, and if it already exists we complain, otherwise set
+%   the \enquote{created} flag for the hook so that it errors next time
+%   \cs{hook_new:n} is used.
 %    \begin{macrocode}
-    \hook_if_exist:nTF {#1}
+    \@@_if_declared:nTF {#1}
       { \msg_error:nnn { hooks } { exists } {#1} }
+      {
+        \tl_new:c { g_@@_#1_declared_tl }
+        \@@_make_usable:n {#1}
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+%  
+%  
+%  
+%  \begin{macro}{\@@_make_usable:n}
+%
+%    This initializes all hook data structures for the hook but if
+%    used on its own doesn mark the hook as declared (as
+%    \cs{hook_new:n} does, so a later\cs{hook_new:n} on that hook will
+%    not result in an error.  This command is internally used by
+%    \cs{hook_gput_code:n} when adding code to a generic hook.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_make_usable:n #1
+  {
 %    \end{macrocode}
-%    Otherwise we add the hook name to the list of all hooks and
-%    allocate the necessary data structures for the new hook.
+%   Now we check if the hook's data structure can be safely created
+%   without \pkg{expl3} raising errors, then
+%   we add the hook name to the list of all hooks and
+%   allocate the necessary data structures for the new hook,
+%   otherwise just do nothing.
+% \fmi{As this internal code we can make sure that this test here is
+%    actually not needed as it was made earlier already -- not changed}    
```

Comment:
well, you could move the test up the chain so that by the end of the day it is always only executed once. But I think you are right it is more natural and safer to test here even if that means one unnecessary test once in a while.